### DISCUSSION TITLE
Various questions about implementation

### DISCUSSION THREAD

**Discussion by cestpasphoto** [suragnair/alpha-zero-general#239] [Q&A]
Hi,

First, thank you very much for all contributors to this repo which is highly pedagogical!
As a newbie in reinforcement learning and machine learning in general, I started to fork your project and during this, it raised some questions to me:

1. Random generation of batch
In [NNet.py](https://github.com/suragnair/alpha-zero-general/blob/master/othello/pytorch/NNet.py#L52), samples for each batch are randomly sampled independently of previous batches of same epoch. Pytorch and Tensorflow implementations share same technic.
But in this [stackoverflow answer](https://stackoverflow.com/a/47504599), training code is different and ensure no common sample between batches inside same epoch. Are both approaches acceptable?

2. Double random?
In [Coach.py](https://github.com/suragnair/alpha-zero-general/blob/master/Coach.py#L106), `trainExamples` is shuffled before calling `nnet.train(trainExamples)`. But what is the point if batches are sampled completely randomly from it?

3. No use of lr variable?
It looks like learning rate parameter `lr` isn't used at all, especially not in [Adam optimizer constructor](https://github.com/suragnair/alpha-zero-general/blob/master/othello/pytorch/NNet.py#L40), same in tensorflow implementation. Did I missed something?

4. Optimizer history
In [NNet.py](https://github.com/suragnair/alpha-zero-general/blob/master/othello/pytorch/NNet.py#L40), Adam optimizer is re-created at each iteration. Wouldn't it be better to keep it and reuse it? Maybe the optimizer has some kind of history and it's better not to start from scratch at each training?

Thank you in advance!
